### PR TITLE
Implemented Permission fix for sys.server_principals

### PR DESCRIPTION
**Pull request by anju15bharti**
**Description**

sys.server_principals view was returning same data irrespective of whether login is a member of sysadmin or not.
Ideally, when we connect with login which is not a member of sysadmin it should only show the rows for current login and fixed_server_role(sysadmin).On the other hand, when we connect with login which is a member of sysadmin it should show all rows.

Task: BABEL-4361
Authored-by: Anju Bharti [abanju@amazon.com](mailto:abanju@amazon.com)
Signed-off-by: Anju Bharti [abanju@amazon.com](mailto:abanju@amazon.com)

**Issues Resolved**
JIRA BABEL-4361
sys.server_principals view was returning same data irrespective of whether login is a member of sysadmin or not.

### Test Scenarios Covered ###
* **Use case based -**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by shalinilohia50 — APPROVED** [review]

File: contrib/babelfishpg_tsql/sql/ownership.sql [hunk 1]
```diff
@@ -323,7 +323,10 @@ CAST(Ext.default_language_name AS SYS.SYSNAME) AS default_language_name,
 CAST(CASE WHEN Ext.type = 'R' THEN NULL ELSE Ext.credential_id END AS INT) AS credential_id,
 CAST(CASE WHEN Ext.type = 'R' THEN 1 ELSE Ext.owning_principal_id END AS INT) AS owning_principal_id,
 CAST(CASE WHEN Ext.type = 'R' THEN 1 ELSE Ext.is_fixed_role END AS sys.BIT) AS is_fixed_role
-FROM pg_catalog.pg_roles AS Base INNER JOIN sys.babelfish_authid_login_ext AS Ext ON Base.rolname = Ext.rolname;
+FROM pg_catalog.pg_roles AS Base INNER JOIN sys.babelfish_authid_login_ext AS Ext ON Base.rolname = Ext.rolname
+WHERE pg_has_role(suser_name()::TEXT, 'sysadmin'::TEXT, 'MEMBER')
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/sql/ownership.sql]: We should pass session_user's oid as first argument directly, it will save a conversion from rolename to roleoid conversion. You can check the `pg_has_role_TYPE_TYPE` family of functions like 'pg_has_role_name_name'.

**Review by HarshLunagariya — CHANGES_REQUESTED** [review]

File: test/JDBC/upgrade/master/schedule [hunk 2]
```diff
@@ -259,7 +259,6 @@ sys-schema-name
 sys-schemas
 sys-schemas-dep
 sys-selective_xml_index_paths
-sys-server_principals
```
Comment (HarshLunagariya) [inline test/JDBC/upgrade/master/schedule:262]: Why this change?

Comment (anju15bharti) [reply test/JDBC/upgrade/master/schedule:262]: I have created logins in prepare testfile and logins are not dumped in case of database-level dump/restore so that's why removed this test from master/schedule file

File: contrib/babelfishpg_tsql/sql/ownership.sql (hunk 1 above)
Comment (anju15bharti) [reply contrib/babelfishpg_tsql/sql/ownership.sql]: done.

File: contrib/babelfishpg_tsql/sql/ownership.sql [hunk 3]
```diff
@@ -323,7 +323,10 @@ CAST(Ext.default_language_name AS SYS.SYSNAME) AS default_language_name,
 CAST(CASE WHEN Ext.type = 'R' THEN NULL ELSE Ext.credential_id END AS INT) AS credential_id,
 CAST(CASE WHEN Ext.type = 'R' THEN 1 ELSE Ext.owning_principal_id END AS INT) AS owning_principal_id,
 CAST(CASE WHEN Ext.type = 'R' THEN 1 ELSE Ext.is_fixed_role END AS sys.BIT) AS is_fixed_role
-FROM pg_catalog.pg_roles AS Base INNER JOIN sys.babelfish_authid_login_ext AS Ext ON Base.rolname = Ext.rolname;
+FROM pg_catalog.pg_roles AS Base INNER JOIN sys.babelfish_authid_login_ext AS Ext ON Base.rolname = Ext.rolname
+WHERE pg_has_role(suser_id(), 'sysadmin'::TEXT, 'MEMBER')
+OR Ext.orig_loginname = suser_name()
+OR Ext.type = 'R';
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_tsql/sql/ownership.sql:329]: Suggestion: the 3rd condition (Ext.type = 'R') might be redundant since the 1st condition will always be true for `sysadmin` role and type = R is only true for `sysadmin`.

**Review by rishabhtanwar29 — APPROVED** [review]
LGTM!

File: contrib/babelfishpg_tsql/sql/ownership.sql (hunk 3 above)
Comment (anju15bharti) [reply contrib/babelfishpg_tsql/sql/ownership.sql:329]: actually not because when login is not a member of sysadmin then we need to return all rows for fixed server role and if we remove the last line it won't return that.

**Review by HarshLunagariya — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]
